### PR TITLE
fix improper determination of 'return this' status

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: test/fail_compilation/retscope.d [hunk 1]
```diff
@@ -349,3 +349,37 @@ int* bar11(scope int** x) @safe
...
+int* escape12() @safe
+{
+    Foo12 f;
+    return f.foo;
+}
+
+struct Foo12
+{
+    int* foo() return @safe;
+}
```
Comment (Geod24) [inline test/fail_compilation/retscope.d]: This is not the test case I provided. The original test case was

```D
int* escape () @safe
{
    int i;
    Foo f;
    f.v = &i;
    return f.foo;
}

struct Foo
{
    int* v;
    int* foo () @safe { return this.v; }
}
```

And still compiles with this P.R.

File: test/fail_compilation/retscope.d [hunk 2]
```diff
@@ -349,3 +349,37 @@ int* bar11(scope int** x) @safe
...
+
+int* escape13() @safe
+{
+    scope Foo13 f;
```
Comment (Geod24) [inline test/fail_compilation/retscope.d]: I would hold off on any `scope class` test before the fix in stable gets merged into scope. @MartinNowak ?

Comment (MartinNowak) [inline test/fail_compilation/retscope.d]: I'll try to get it here.

**Comment by WalterBright** [comment]
This is very strange. The error produced is compiling Phobos unittests:
```
std\typecons.d(5374): Error: circular typeof definition
std\typecons.d(5374): Error: alias std.typecons.__unittestL5988_127.S.Proxy!(foo).ValueType recursive alias declaration
std\typecons.d(5994): Error: mixin std.typecons.__unittestL5988_127.S.Proxy!(foo) error instantiating\
```
which does not change even if I disable the changes in this PR.

**Comment by WalterBright** [comment]
https://github.com/dlang/dmd/pull/6331 seems to be the culprit.

**Comment by WalterBright** [comment]
@MartinNowak ready to rock!

File: src/escape.d [hunk 3]
```diff
@@ -865,7 +867,16 @@ private void escapeByValue(Expression e, EscapeByResults* er)
             if (e.e1.op == TOKdotvar && t1.ty == Tfunction)
             {
                 DotVarExp dve = cast(DotVarExp)e.e1;
-                if (dve.var.storage_class & STCreturn || tf.isreturn)
+                FuncDeclaration fd = dve.var.isFuncDeclaration();
+                AggregateDeclaration ad;
+                if (global.params.vsafe && tf.isreturn && fd && (ad = fd.isThis()) !is null)
+                {
+                    if (ad.isClassDeclaration())       // this is 'return scope'
+                        dve.e1.accept(this);
```
Comment (MartinNowak) [inline src/escape.d]: Coverage missing, please add a test case.
Also see https://github.com/codecov/browser-extension.

File: test/fail_compilation/retscope.d [hunk 4]
```diff
@@ -352,25 +352,13 @@ int* foo11(int* x) @safe { return x; }
...
 TEST_OUTPUT:
-// FIXME FIXME FIXME
-fail_compilation/retscope.d(1003): Error: escaping reference to local variable f
-// FIXME FIXME FIXME
 ---
+fail_compilation/retscope.d(365): Error: escaping reference to local variable f
 ---
 */
 
-#line 1000
```
Comment (MartinNowak) [inline test/fail_compilation/retscope.d]: It rather seems like a good idea to make tests position independent, what's the problem with it?

File: test/fail_compilation/retscope.d [hunk 5]
```diff
@@ -424,3 +413,26 @@ struct Foo14
...
+/*
+TEST_OUTPUT:
+// FIXME FIXME FIXME
+fail_compilation/retscope.d(1003): Error: escaping reference to local variable f
+// FIXME FIXME FIXME
+---
+---
+*/
+
+#line 1000
```
Comment (MartinNowak) [inline test/fail_compilation/retscope.d]: Should be deleted now if you fixed the test.

**Review by MartinNowak — APPROVED** [review]
I'll do another rebase, things got messed up a bit.

File: src/escape.d (hunk 3 above)
Comment (andralex) [inline src/escape.d]: @WalterBright you may want to install the CodeCov Chrome extension, it nicely shows uncovered lines in red on this page!

**Review by andralex — APPROVED** [review]
nice!

**Comment by MartinNowak** [comment]
Just removed the noise from the weird rebase, now it's clearer what this PR actually does.
Still please spend more than 2 seconds writing commit messages and PR descriptions, saves a lot of guesswork on the other side.
"fix improper determination of status" conveys almost no information.

**Comment by MartinNowak** [comment]
Don't try this at home kids, we're still testing our new auto-merge feature.
https://github.com/MartinNowak/dlang-bot/pull/17

**Comment by andralex** [comment]
@MartinNowak COOL COOL COOL!

**Pull request merged**